proposals/0072-feature-gate-threshold-automation.md

```diff
@@ -171,15 +171,17 @@ A node's submitted bit mask is then used to add that node's current epoch stake
 to the Staged Features PDA for each feature for which it has signaled support.
 This is done by using the Get Epoch Stake Syscall.
 
-Nodes should send a transaction containing this instruction at some arbitrary
-point during the epoch at least 128 slots before the end of the epoch and on
-startup after any reboot. Transactions sent too late (< 128 slots before epoch
-end) will be rejected by the Feature Gate program.
-
-If a node does not send this transaction, or sends it too late in the epoch (>
-128 slots before the end), their stake is not tallied. This is analogous to a
-node sending this transaction in a valid point in the epoch signalling support
-for zero features.
+Nodes are required to submit a transaction containing this instruction:
+
+- Any time at least 4500 slots before the end of the epoch.
+- During startup after any reboot.
+
+Transactions sent too late (< 4500 slots before the epoch end) will be rejected
+by the Feature Gate program.
```

Comment:
Why do we need any deadline/restriction? I see the program cache being alluded to, but I don't see any specification that changes to the Staged Features PDA must begin or conclude at some point earlier than the epoch boundary. It's not obvious why the Feature Gate program can't continue to modify the Staged Features PDA up to the epoch boundary.

Reply:
Are you suggesting to explicitly explain the reasoning for this constraint in the SIMD, or are you asking why we actually need such a constraint? In either case, I agree it should be more explicit in the proposal.

It seems to me that @Lichtso intends to increase the cache preparation phase's initiation to be sooner in the epoch ([comment](https://github.com/solana-foundation/solana-improvement-documents/pull/72#discussion_r1603757414)), so we probably need a more concrete explanation as to why the Feature Gate program should have this constraint.

In this proposal, it was initially set to 128 slots remaining in the epoch because the cache preparation stage begins at a maximum of 128 slots before the end of the epoch, [as computed here](https://github.com/anza-xyz/agave/blob/9dd01f55182ff75a9c21bf89958ba5b37c8fe937/svm/src/transaction_processor.rs#L478-L482).

It's worth calling out that if we configure this constraint in the Feature Gate program and then later need to adjust it for an earlier cache preparation start, we'll be bound to using a feature gate for that change to the program.

Reply:
I'm asking why we need a constraint at all. The cache being referenced is supposedly an agave-specific implementation, not part of the runtime protocol. If feature activation needs to work around any aspect of cache behavior, that seems like (a) a problem; and (b) a thing that definitely needs to be specified for all clients.

Reply:
There's really nothing stopping the cache from being able to anticipate the upcoming feature environment by querying the PDA under the Feature Gate program that would tell it what the current stake support is for each feature. It's actually arguably equally or more deterministic than the way it's done now.

The cache could run the calculation at it's current 128 slots before epoch boundary and probably get a pretty good understanding of which features have enough stake support. Increasing that slot buffer beyond 128 could just mean increasing the likelihood that not enough stake has been signaled for a feature. Perhaps this is a compromise the cache implementation should consider? After all, I believe recompilation is an optimization.

Reply:
I've added a commit removing the "slots remaining" constraint, since I think the above is a sensible direction to head in. If there's explicit opposition, it can be added back in.

Reply:
@CriesofCarrots This is not only for the program cache and not only for our implementation. All validators must reach consensus before feature activation. If a feature is activated to close to the epoch boundary it is possible that multiple forks, some with the feature active others with it inactive, cross the boundary independently.

Reply:
Features are always activated exactly on the epoch boundary, now and in this SIMD. I think maybe you mean something other than what you said?

Reply:
Sorry for my imprecise formulation, I meant scheduled for activation.
If a feature is scheduled for activation in some slot close to the boundary, then it is possible that the TX is not yet rooted. Thus, some validator nodes will not have it scheduled, and will not activate it on the boundary.

Reply:
Please explain how this is different from the cluster coming to consensus on other tx/block.

Reply:
For features which only affect their own fork this is irrelevant. But for features which change validator global behavior this would break. And I am not sure if there is much to be gained from making exceptions or harving different feature types.

Reply:
Just adding an additional point of clarity here, if it helps.

Right now, in Agave, the cache is calling `compute_active_feature_set`, which will compute the feature set based on the on-chain accounts that have been created under `Feature11111...` (ie. "activated"). The cache does this ~128 slots before the epoch boundary to prepare the cache entries for the anticipated upcoming environment.

With the proposed system in place, the cache can still call `compute_active_feature_set` the same way it does now. The function will just compute the anticipated feature set from the proposed PDA (with stake support) rather than just the presence of on-chain accounts.

In almost every way it's the same as what we have now. The subtle difference is that someone can no longer activate a feature after the cache has done its feature-set computation, but nodes could still cast votes (signals) on a feature, bringing it to "eligible" status. This edge case exists today and just causes cache preparation/recompilation to be moot.

What @CriesofCarrots is suggesting is that we don't explicitly architect this mechanism around the cache, which is sensible, especially considering the cache preparation may change.

Reply:
@Lichtso do you still have concerns here? I'd like to move this SIMD along.